# Assignment 1 - Build a performing machine Regression algorithm

### Objective:
The objective of the assignment is to develop a linear regression machine learning model

### Data Description:
We have collected data on students' performance from UCI data respository. The data shows peformance of students in Gabriel Pereira and Mousinho da Silveira schools. The dataset has the following features:

### Attributes of the dataset that shows students performance in math course 
1. school - student's school (binary: 'GP' - Gabriel Pereira or 'MS' - Mousinho da Silveira) 
2. sex - student's sex (binary: 'F' - female or 'M' - male) 
3. age - student's age (numeric: from 15 to 22) 
4. address - student's home address type (binary: 'U' - urban or 'R' - rural) 
5. famsize - family size (binary: 'LE3' - less or equal to 3 or 'GT3' - greater than 3) 
6. Pstatus - parent's cohabitation status (binary: 'T' - living together or 'A' - apart) 
7. Medu - mother's education (numeric: 0 - none, 1 - primary education (4th grade), 2 – 5th to 9th grade, 3 – secondary education or 4 – higher education) 
8. Fedu - father's education (numeric: 0 - none, 1 - primary education (4th grade), 2 – 5th to 9th grade, 3 – secondary education or 4 – higher education) 
9. Mjob - mother's job (nominal: 'teacher', 'health' care related, civil 'services' (e.g. administrative or police), 'at_home' or 'other') 
10. Fjob - father's job (nominal: 'teacher', 'health' care related, civil 'services' (e.g. administrative or police), 'at_home' or 'other') 
11. reason - reason to choose this school (nominal: close to 'home', school 'reputation', 'course' preference or 'other') 
12. guardian - student's guardian (nominal: 'mother', 'father' or 'other') 
13. traveltime - home to school travel time (numeric: 1 - <15 min., 2 - 15 to 30 min., 3 - 30 min. to 1 hour, or 4 - >1 hour) 
14. studytime - weekly study time (numeric: 1 - <2 hours, 2 - 2 to 5 hours, 3 - 5 to 10 hours, or 4 - >10 hours) 
15. failures - number of past class failures (numeric: n if 1<=n<3, else 4) 
16. schoolsup - extra educational support (binary: yes or no) 
17. famsup - family educational support (binary: yes or no) 
18. paid - extra paid classes within the course subject (Math or Portuguese) (binary: yes or no) 
19. activities - extra-curricular activities (binary: yes or no) 
20. nursery - attended nursery school (binary: yes or no) 
21. higher - wants to take higher education (binary: yes or no) 
22. internet - Internet access at home (binary: yes or no) 
23. romantic - with a romantic relationship (binary: yes or no) 
24. famrel - quality of family relationships (numeric: from 1 - very bad to 5 - excellent) 
25. freetime - free time after school (numeric: from 1 - very low to 5 - very high) 
26. goout - going out with friends (numeric: from 1 - very low to 5 - very high) 
27. Dalc - workday alcohol consumption (numeric: from 1 - very low to 5 - very high) 
28. Walc - weekend alcohol consumption (numeric: from 1 - very low to 5 - very high) 
29. health - current health status (numeric: from 1 - very bad to 5 - very good) 
30. absences - number of school absences (numeric: from 0 to 93) 

### These are grades secured by students at the end of each period
31. G1 - first period grade (numeric: from 0 to 20) 
31. G2 - second period grade (numeric: from 0 to 20) 
32. G3 - final grade (numeric: from 0 to 20, output target)

In [160]:
# import essential libraries
import pandas as pd
import numpy as np
pd.set_option("display.max_columns", None)
# adjust screensize
from IPython.core.display import display, HTML
display(HTML("<style>.container { width:100% !important; }</style>"))

/var/folders/n1/gcdj6fzx2vl4c14b35_8tbzh0000gn/T/ipykernel_71005/922315407.py:6: DeprecationWarning: Importing display from IPython.core.display is deprecated since IPython 7.14, please import from IPython display
  from IPython.core.display import display, HTML


In [161]:
mdata = pd.read_csv("student-mat.csv", sep=";")
mdata.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

   reason guardian  traveltime  studytime  failures schoolsup famsup paid  \
0  course   mother           2          2         0       yes     no   no   
1  course   father           1          2         0        no    yes   no   
2   other   mother           1          2         3       yes     no  yes   
3    home   mother           1          3         0        no    yes  yes   
4    home   father           1          2         0        no    yes  yes   

  activities nursery higher internet romantic  famrel  freetime  goout  Dalc  \
0         no     yes    yes       no       no       4         3      4     1   
1         no      no    yes      yes       no       5         3      3     1   
2         no     yes    yes      yes       no       4         3      2     2   
3        yes     yes    yes      yes      yes       3         2      2     1   
4         no     yes    yes       no       no       4         3      2     1   

   Walc  health  absences  G1  G2  G3  
0     1       3         6   5   6   6  
1     1       3         4   5   5   6  
2     3       3        10   7   8  10  
3     1       5         2  15  14  15  
4     2       5         4   6  10  10

In [162]:
mdata.info() # identifying the categorical varibles and checking for null values

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 33 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   school      395 non-null    object
 1   sex         395 non-null    object
 2   age         395 non-null    int64 
 3   address     395 non-null    object
 4   famsize     395 non-null    object
 5   Pstatus     395 non-null    object
 6   Medu        395 non-null    int64 
 7   Fedu        395 non-null    int64 
 8   Mjob        395 non-null    object
 9   Fjob        395 non-null    object
 10  reason      395 non-null    object
 11  guardian    395 non-null    object
 12  traveltime  395 non-null    int64 
 13  studytime   395 non-null    int64 
 14  failures    395 non-null    int64 
 15  schoolsup   395 non-null    object
 16  famsup      395 non-null    object
 17  paid        395 non-null    object
 18  activities  395 non-null    object
 19  nursery     395 non-null    object
 20  higher    

In [163]:
mdata.shape

(395, 33)

As one can see above, there is a total of 395 samples of data with 33 features. Fortunately, none of the features has any missing values, so there is no need to either drop missing values or to impute them. However, we notice that many features are stored as objects (categorical values). As a result, we need to convert these values to numbers so that the model can interpret them.

In [164]:
mdata.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

   reason guardian  traveltime  studytime  failures schoolsup famsup paid  \
0  course   mother           2          2         0       yes     no   no   
1  course   father           1          2         0        no    yes   no   
2   other   mother           1          2         3       yes     no  yes   
3    home   mother           1          3         0        no    yes  yes   
4    home   father           1          2         0        no    yes  yes   

  activities nursery higher internet romantic  famrel  freetime  goout  Dalc  \
0         no     yes    yes       no       no       4         3      4     1   
1         no      no    yes      yes       no       5         3      3     1   
2         no     yes    yes      yes       no       4         3      2     2   
3        yes     yes    yes      yes      yes       3         2      2     1   
4         no     yes    yes       no       no       4         3      2     1   

   Walc  health  absences  G1  G2  G3  
0     1       3         6   5   6   6  
1     1       3         4   5   5   6  
2     3       3        10   7   8  10  
3     1       5         2  15  14  15  
4     2       5         4   6  10  10

In [165]:
for i in mdata.columns:
    print(f"Values in {i}: {mdata[i].unique()}")

Values in school: ['GP' 'MS']
Values in sex: ['F' 'M']
Values in age: [18 17 15 16 19 22 20 21]
Values in address: ['U' 'R']
Values in famsize: ['GT3' 'LE3']
Values in Pstatus: ['A' 'T']
Values in Medu: [4 1 3 2 0]
Values in Fedu: [4 1 2 3 0]
Values in Mjob: ['at_home' 'health' 'other' 'services' 'teacher']
Values in Fjob: ['teacher' 'other' 'services' 'health' 'at_home']
Values in reason: ['course' 'other' 'home' 'reputation']
Values in guardian: ['mother' 'father' 'other']
Values in traveltime: [2 1 3 4]
Values in studytime: [2 3 1 4]
Values in failures: [0 3 2 1]
Values in schoolsup: ['yes' 'no']
Values in famsup: ['no' 'yes']
Values in paid: ['no' 'yes']
Values in activities: ['no' 'yes']
Values in nursery: ['yes' 'no']
Values in higher: ['yes' 'no']
Values in internet: ['no' 'yes']
Values in romantic: ['no' 'yes']
Values in famrel: [4 5 3 1 2]
Values in freetime: [3 2 4 1 5]
Values in goout: [4 3 2 1 5]
Values in Dalc: [1 2 5 3 4]
Values in Walc: [1 3 2 4 5]
Values in health: [3 5

The above view of data enables us to determine which features need to be translated into dummy variables. 

In [166]:
# a function created to convert categorical features into dummy variables.
def dummy_creator(arr, pre):
    arr_dummy = pd.get_dummies(arr, prefix=pre, drop_first=True) # dropped the first dummy variable to avoid dummy variable trap
    return arr_dummy

In [167]:
mdata_categorical = mdata.drop(mdata.select_dtypes([np.number]), axis=1) # to extract all categorical data

In [168]:
for col in mdata_categorical.columns:
    mdata_categorical = pd.concat([mdata_categorical, dummy_creator(mdata_categorical[col], col)], axis=1) # converts features into dummy variables and concatenate them to the dataframe
mdata_categorical.head()

school sex address famsize Pstatus     Mjob      Fjob  reason guardian  \
0     GP   F       U     GT3       A  at_home   teacher  course   mother   
1     GP   F       U     GT3       T  at_home     other  course   father   
2     GP   F       U     LE3       T  at_home     other   other   mother   
3     GP   F       U     GT3       T   health  services    home   mother   
4     GP   F       U     GT3       T    other     other    home   father   

  schoolsup famsup paid activities nursery higher internet romantic  \
0       yes     no   no         no     yes    yes       no       no   
1        no    yes   no         no      no    yes      yes       no   
2       yes     no  yes         no     yes    yes      yes       no   
3        no    yes  yes        yes     yes    yes      yes      yes   
4        no    yes  yes         no     yes    yes       no       no   

   school_MS  sex_M  address_U  famsize_LE3  Pstatus_T  Mjob_health  \
0          0      0          1            0          0            0   
1          0      0          1            0          1            0   
2          0      0          1            1          1            0   
3          0      0          1            0          1            1   
4          0      0          1            0          1            0   

   Mjob_other  Mjob_services  Mjob_teacher  Fjob_health  Fjob_other  \
0           0              0             0            0           0   
1           0              0             0            0           1   
2           0              0             0            0           1   
3           0              0             0            0           0   
4           1              0             0            0           1   

   Fjob_services  Fjob_teacher  reason_home  reason_other  reason_reputation  \
0              0             1            0             0                  0   
1              0             0            0             0                  0   
2              0             0            0             1                  0   
3              1             0            1             0                  0   
4              0             0            1             0                  0   

   guardian_mother  guardian_other  schoolsup_yes  famsup_yes  paid_yes  \
0                1               0              1           0         0   
1                0               0              0           1         0   
2                1               0              1           0         1   
3                1               0              0           1         1   
4                0               0              0           1         1   

   activities_yes  nursery_yes  higher_yes  internet_yes  romantic_yes  
0               0            1           1             0             0  
1               0            0           1             1             0  
2               0            1           1             1             0  
3               1            1           1             1             1  
4               0            1           1             0             0

We notice above that all the categorical features have been converted to dummy values and concatenated to the categorical dataset

In [169]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler() # importing MinMax Scaler to scale numerical data within the range of [0,1] so that they are consistent with the categorical ones. This will ensure 
# that data with greater magnitude will not sway the model results towards itself. 

In [170]:
mdata_continuous = mdata._get_numeric_data() # extracting all the numeric data from the original dataset
mdata_cont_scaled = pd.DataFrame(scaler.fit_transform(mdata_continuous), columns=mdata_continuous.columns) # scaled the continuous variables so that they are in the 
# same range as the categoricla ones
mdata_cont_scaled.head()

age  Medu  Fedu  traveltime  studytime  failures  famrel  freetime  \
0  0.428571  1.00  1.00    0.333333   0.333333       0.0    0.75      0.50   
1  0.285714  0.25  0.25    0.000000   0.333333       0.0    1.00      0.50   
2  0.000000  0.25  0.25    0.000000   0.333333       1.0    0.75      0.50   
3  0.000000  1.00  0.50    0.000000   0.666667       0.0    0.50      0.25   
4  0.142857  0.75  0.75    0.000000   0.333333       0.0    0.75      0.50   

   goout  Dalc  Walc  health  absences      G1        G2    G3  
0   0.75  0.00  0.00     0.5  0.080000  0.1250  0.315789  0.30  
1   0.50  0.00  0.00     0.5  0.053333  0.1250  0.263158  0.30  
2   0.25  0.25  0.50     0.5  0.133333  0.2500  0.421053  0.50  
3   0.25  0.00  0.00     1.0  0.026667  0.7500  0.736842  0.75  
4   0.25  0.00  0.25     1.0  0.053333  0.1875  0.526316  0.50

In [171]:
mdata_new = pd.concat([mdata_categorical, mdata_cont_scaled], axis=1) # merged the numerical data with categorical ones
mdata_new = mdata_new._get_numeric_data() # since categorical data has been converted to dummy variables, we now remove the original categorical data with string values
mdata_new

school_MS  sex_M  address_U  famsize_LE3  Pstatus_T  Mjob_health  \
0            0      0          1            0          0            0   
1            0      0          1            0          1            0   
2            0      0          1            1          1            0   
3            0      0          1            0          1            1   
4            0      0          1            0          1            0   
..         ...    ...        ...          ...        ...          ...   
390          1      1          1            1          0            0   
391          1      1          1            1          1            0   
392          1      1          0            0          1            0   
393          1      1          0            1          1            0   
394          1      1          1            1          1            0   

     Mjob_other  Mjob_services  Mjob_teacher  Fjob_health  Fjob_other  \
0             0              0             0            0           0   
1             0              0             0            0           1   
2             0              0             0            0           1   
3             0              0             0            0           0   
4             1              0             0            0           1   
..          ...            ...           ...          ...         ...   
390           0              1             0            0           0   
391           0              1             0            0           0   
392           1              0             0            0           1   
393           0              1             0            0           1   
394           1              0             0            0           0   

     Fjob_services  Fjob_teacher  reason_home  reason_other  \
0                0             1            0             0   
1                0             0            0             0   
2                0             0            0             1   
3                1             0            1             0   
4                0             0            1             0   
..             ...           ...          ...           ...   
390              1             0            0             0   
391              1             0            0             0   
392              0             0            0             0   
393              0             0            0             0   
394              0             0            0             0   

     reason_reputation  guardian_mother  guardian_other  schoolsup_yes  \
0                    0                1               0              1   
1                    0                0               0              0   
2                    0                1               0              1   
3                    0                1               0              0   
4                    0                0               0              0   
..                 ...              ...             ...            ...   
390                  0                0               1              0   
391                  0                1               0              0   
392                  0                0               1              0   
393                  0                1               0              0   
394                  0                0               0              0   

     famsup_yes  paid_yes  activities_yes  nursery_yes  higher_yes  \
0             0         0               0            1           1   
1             1         0               0            0           1   
2             0         1               0            1           1   
3             1         1               1            1           1   
4             1         1               0            1           1   
..          ...       ...             ...          ...         ...   
390           1         1               0            1           1   
391           0         0               0    

In [172]:
# We split the dataset into dependent (Final grade-G3) and independent variables (all the remaining features)
# Given our lack of domain expertise, we have not eliminated any feature yet. We intend to do that later in the analysis based on their effect on the final score.
x = mdata_new.iloc[:, :-1]
y = mdata_new.iloc[:, -1]

In [173]:
# import train test split package from scikit learn library
from sklearn.model_selection import train_test_split

In [174]:
# We are dividing the dataset into training and testing splits. We retain 20% of the data for testing purpose
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

In [175]:
from sklearn.linear_model import LinearRegression

In [176]:
model = LinearRegression() # creating a linear regression object

In [177]:
model.fit(x_train, y_train)

LinearRegression()

In [178]:
from sklearn.metrics import r2_score

In [179]:
y_predicted = model.predict(x_test)

In [180]:
score_wo_RFE = r2_score(y_predicted, y_test) # this score is based on all the features prior to applying any form of feature selection
score_wo_RFE

0.7102320490925691

In [181]:
from sklearn.feature_selection import RFE # importing Recursive Feature Elimination feature from Scikit Learn Library

In [182]:
feature_num = list()
scores = list()
for index in range(1,41):
    sel = RFE(LinearRegression(), n_features_to_select=index) # we use a loop to iterate through the number of features starting from 1 to 41 (the total number of features)
    sel.fit(x_train, y_train)
    x_train_rfe = sel.transform(x_train) # RFE is applied on training set based on the number of features selected iteratively
    x_test_rfe = sel.transform(x_test) # RFE is applied on test data based on the number of features selected iteratively
    model = LinearRegression() 
    model.fit(x_train_rfe, y_train)
    y_predicted = model.predict(x_test_rfe)
    feature_num.append(index)
    scores.append(r2_score(y_predicted, y_test)) # predicting r2 scores for combination of features from 1 to 41

score_dict = dict(zip(feature_num, scores))
score_dict # a dictionary that shows number of features as keys and accuracy scores as values

{1: 0.7595167048116915,
 2: 0.7556024757151003,
 3: 0.770429722878803,
 4: 0.7731287745087871,
 5: 0.7745936863734232,
 6: 0.7629448118384672,
 7: 0.75218904060062,
 8: 0.7452886904816265,
 9: 0.7386850045269882,
 10: 0.7298416554465669,
 11: 0.7243661440810729,
 12: 0.7217916106222884,
 13: 0.7223742626227214,
 14: 0.7268821319734143,
 15: 0.7263938877853681,
 16: 0.7274509119154694,
 17: 0.719658942084907,
 18: 0.7203266271850115,
 19: 0.7184690754794398,
 20: 0.7165867624916803,
 21: 0.7147668020119822,
 22: 0.7147689339529146,
 23: 0.7109168365676874,
 24: 0.712111820706428,
 25: 0.7091867745472341,
 26: 0.7090021270929527,
 27: 0.7091738274543716,
 28: 0.7094938997206504,
 29: 0.7103012794527541,
 30: 0.7110552547495368,
 31: 0.7100133214229583,
 32: 0.7095480296900616,
 33: 0.7101530020211748,
 34: 0.7113216939208233,
 35: 0.7123262633526639,
 36: 0.711505245630137,
 37: 0.7111562254275208,
 38: 0.7102886030751937,
 39: 0.7100218246676416,
 40: 0.7102845072992834}

In [183]:
max_key = max(score_dict, key=score_dict.get)
max_key # show that RFE has created a combination of five features that produces the maximum accuracy score

5

### Understanding the Features and their implications
Since we have already discovered that RFE provides the highest scores when five features are selected, we run the model with 5 features and note that results below.

In [184]:
select = RFE(LinearRegression(), n_features_to_select=5)

In [185]:
select.fit(x_train, y_train)
x_train_selected = select.transform(x_train)
x_test_selected = select.transform(x_test)

In [186]:
model_new = LinearRegression()

In [187]:
model_new.fit(x_train_selected, y_train)

LinearRegression()

In [188]:
y_predicted_new = model_new.predict(x_test_selected)

In [192]:
score_with_RFE = r2_score(y_predicted_new, y_test)
score_with_RFE

0.7745936863734232

In [198]:
x_train.columns[select.get_support()]
# Based on RFE we discover that age, quality of family relationships, number of absenses, scores in first and second periods are the most influencial features when predicting 
# final scores of students in mathematics.

Index(['age', 'famrel', 'absences', 'G1', 'G2'], dtype='object')

In [197]:
# change in score
"{:.2f}%".format((score_with_RFE - score_wo_RFE)/score_wo_RFE * 100)

'9.06%'

### Conclusion:
We set out to create a machine learning linear regression model that predicts final score of students in mathematics. We started with 32 features and got an R2 score of 0.71. Given any R2 score above 0.6 signifies a good model, it is safe to assume that our model is predicting scores well. However, we wanted to enhance the effectiveness and efficiency of the model by recursively eliminating features and assessing the impact on the accuracy score. 

Using the Recursive Feature Elimination package of Scikit Learn Library, we discovered that the best R2 score is achieved when we limit the number of features to five (age, quality of family relationship, number of absences, score in first period and score in second period). By using RFE, we managed to increase the accuracy by more than 9 percentage points and reduce the number of features from 32 to 5. In addition to providing more reliable predictions, we believe application of RFE will significantly reduce strain on computational resources.